text/0000-declarative-nonstandard-event-binding.md

```diff
@@ -0,0 +1,125 @@
+---
+title: Declarative Binding for Non-Standard Event Names
+status: DRAFTED
+created_at: 2024-02-25
+updated_at: 2024-02-25
+pr: (leave this empty until the PR is created)
+---
+
+# Declarative Binding for Non-Standard Event Names
+
+## Summary
+
+LWC support for declarative event binding has always been limited to event names using lower-case
+alphabetic characters, underscores, and numeric characters. Event names containing upper-case
+characters (e.g., camelCase, CAPScase, PascalCase, etc) have never been supported in LWC due to
+the case-insensitive nature of HTML. In addition, event names containing hyphens were never
+supported due to the reluctance to avoid introducing ambiguity into the LWC convention of using
+hyphens in attribute names to map to upper-case letters in property names.
+
+This has worked well in practice with developers sticking to LWC conventions and using
+`addEventListener()` when required as a workaround, but with the recent introduction of
+`lwc:external` which provides a paved path for integrating third party web components, there is
+greater value in providing a declarative way to listen for non-standard event names.
+
+This RFC proposes the introduction of a new template directive `lwc:on`, which allows us to support
+declarative event binding for third party web components, while sidestepping all the aforementioned
+limitations and ambiguities.
+
+## Basic example
+
+```js
+customElements.define('third-party', class extends HTMLElement{
+    connectedCallback() {
+        this.dispatchEvent(new CustomEvent('Foo-Bar-BAZ'));
+    }
+});
+```
+
+These are examples of some of the options that are currently on the table:
+
+```html
+<!-- Event as part of directive -->
+<third-party lwc:external lwc:on:Foo-Bar-BAZ={handleConnected}></third-party>
+
+<!-- Event as part of attribute name (special treatment for lwc:external) -->
+<third-party lwc:external onFoo-Bar-BAZ={handleConnected}></third-party>

```

Comment:
To me, this option is the least disruptive. Yes, it is playing fast-and-loose with HTML semantics, but it is the most similar to what we have today, avoiding the need for a special new `lwc:on` directive or other mechanism.

Scoping it to `lwc:external` also brings that advantage that we can pass 100% on Custom Elements Everywhere without changing how LWC components have worked up until today. (We can re-evaluate that later if needed, but I really think it's not a big deal – the important thing is just to support the wide world of third-party custom elements and their wacky naming conventions for events.)

Reply:
I agree that it's the least disruptive, but I also feel like it might be a source of confusion/surprise. The nice thing about introducing a new directive is that it's easy to understand...if you want to do this thing that you were never able to before, you have to use this new directive.